// @ndn/fw/lib/tap-face_browser.js
import { __importDefault, __importStar } from "tslib";
import { assert, MultiMap } from "@ndn/util";
import { pushable } from "it-pushable";
import _cjsDefaultImport0 from "mnemonist/default-weak-map.js"; const DefaultWeakMap = __importDefault(_cjsDefaultImport0).default;
import { Forwarder } from "./forwarder_browser.js";
class TapRxController {
    fw;
    static instances = new DefaultWeakMap((fw) => new TapRxController(fw));
    static lookup(fw) {
        return TapRxController.instances.get(fw);
    }
    taps = new MultiMap();
    constructor(fw) {
        this.fw = fw;
        this.fw.on("pktrx", this.pktrx);
        this.fw.on("facerm", this.facerm);
    }
    add(src, dst) {
        assert.equal(src.fw, this.fw);
        this.taps.add(src, dst);
    }
    remove(src, dst) {
        this.taps.remove(src, dst);
        this.detachIfIdle();
    }
    facerm = (src) => {
        const dst = this.taps.list(src);
        for (const { rx } of dst) {
            rx.end();
        }
        this.detachIfIdle();
    };
    detachIfIdle() {
        if (this.taps.size === 0) {
            this.fw.off("pktrx", this.pktrx);
            this.fw.off("facerm", this.facerm);
            TapRxController.instances.delete(this.fw);
        }
    }
    pktrx = (src, pkt) => {
        const dst = this.taps.list(src);
        for (const { rx } of dst) {
            rx.push(pkt);
        }
    };
}
/**
 * Create a secondary face by tapping on a primary face.
 *
 * TapFace is useful for sending in-band management commands to a specific neighbor, after being
 * added to a temporary secondary Forwarder. The TapFace shares the same transport as the primary
 * face, but allows independent FIB and PIT settings. The primary Forwarder will see RX packets,
 * but does not see TX packets.
 */
export class TapFace {
    face;
    get attributes() {
        return {
            describe: `tap(${this.face})`,
            ...this.face.attributes,
        };
    }
    rx = pushable({ objectMode: true });
    ctrl;
    constructor(face) {
        this.face = face;
        this.ctrl = TapRxController.lookup(face.fw);
        this.ctrl.add(this.face, this);
    }
    tx = async (iterable) => {
        for await (const pkt of iterable) {
            this.face.send(pkt);
        }
        this.ctrl.remove(this.face, this);
    };
}
(function (TapFace) {
    /** Create a new Forwarder and add a TapFace. */
    function create(face) {
        const fw = Forwarder.create();
        return fw.addFace(new TapFace(face));
    }
    TapFace.create = create;
})(TapFace || (TapFace = {}));
